ERCS/erc-post_execution_4337_validation.md

```diff
@@ -0,0 +1,127 @@
+        }
+
+        // Perform post-bundle execution validation for each userOp
+        for (uint256 i = 0; i < opslen; i++) {
+            _validatePostExecution(ops[i], opInfos[i]);
+        }
+
+        _compensate(beneficiary, collected);
+    }
+    ...
+}
+```
+
+## Rationale
+
+This enhancement aims to unlock new capabilities and scenarios in DeFi, DApps, and beyond by enabling post-execution bundle validation. This feature is particularly beneficial in complex transaction sequences where one operation's outcome may affect subsequent ones' within the same transaction bundle. By allowing for validation after the execution of all bundled operations, this proposal accommodates the requirement for a generic, post-bundle execution validation mechanism that can address an array of conditions, including but not limited to the enforcement of transactional invariants and more sophisticated transaction constructs such as Intents fulfillment, and enhances the bundle's security guarantees.
+
+Scenario Examples
+
+DeFi Sequences 
+A user performs a token swap followed by liquidity provision and staking in one transaction. Post-execution validation ensures that the entire sequence meets the user's conditions based on the final state, such as minimum received tokens.
+
+Blockchain-based Games
+In-game transactions that depend on specific game states can be validated post-execution. Such validation ensures that actions like trades or character upgrades only proceed if prior operations result in the expected state.
+
+Social Recovery
+Validates the completion of a social recovery process based on approvals from designated guardians, ensuring the integrity of the recovery operation post-execution.
+
+Alternative design considerations
+
+During the design phase, we considered the existing `postOp` function utilized by Paymasters as an alternative mechanism for post-operation validation. As defined within the ERC-4337 standard, Paymaster's approach facilitates gas payment delegation and handles validation immediately after executing each user operation. While functionally significant, this approach cannot inherently assess the state changes produced by the entirety of a user operation bundle. Moreover, the semantics and primary intent of the Paymaster's post-operation validation are focused on the financial aspect of operations.
+
+## Backwards Compatibility
+
+Given that each smart account is inherently bound to a specific EntryPoint contract, the ERC's updates do not impose any disruptive changes on existing accounts. Smart accounts that do not implement the new post-execution validation method will continue to operate as before without altering to their transactional behavior or interaction with the EntryPoint. This design choice is pivotal in ensuring that the integration of the new validation mechanism is non-intrusive and preserves the integrity of existing smart account deployments.
+
+## Security Considerations
```

Comment:
I think griefing attack is more accurate than DDos

Reply:
when I read the spec I felt the same which alludes to the susceptibility to DOS attacks 5 times, e.g this instance:
> how can a block builder including an operation make sure that it will actually pay fees, without having to first execute the entire operation? Requiring the block builder to execute the entire operation opens a DoS attack vector, as an attacker could easily send many operations that pretend to pay a fee but then revert at the last moment after a long execution.

I'll change it.